File: C(++)/034. Search for a Range.cpp
/*
034. Search for a Range
https://leetcode.com/problems/search-for-a-range/

cpp 9 ms
*/

/*
Given an array of integers sorted in ascending order, find the starting and ending position of a given target value.

Your algorithm's runtime complexity must be in the order of O(log n).

If the target is not found in the array, return [-1, -1].

For example,
Given [5, 7, 7, 8, 8, 10] and target value 8,
return [3, 4].
*/

class Solution {
public:
    vector<int> n;
    int tg;
    int INF=0x7fffffff;
    int imin(int a,int b)
    {
        return a<b?a:b;
    }
    int imax(int a,int b)
    {
        return a>b?a:b;
    }
    void go(int l,int r,int *al,int *ar)
    {
        while (l<r)
        {
            int m=(l+r)/2,v=n[m];
            if (v>tg)
                r--;
            else if (v<tg)
                l++;
            else
            {
                *al=*ar=m;
                int tal,tar;
                if (m-1>=l&&n[m-1]==tg)
                {
                    go(l,m-1,&tal,&tar);
                    *al=imin(*al,tal);
                    *ar=imax(*ar,tar);
                }
                if (m+1<=r&&n[m+1]==tg)
                {
                    go(m+1,r,&tal,&tar);
                    *al=imin(*al,tal);
                    *ar=imax(*ar,tar);
                }
                break;
            }
        }
        if (l==r&&n[l]==tg)
            *al=*ar=l;
    }
    vector<int> searchRange(vector<int>& nums, int target) {
        n=nums;
        tg=target;
        int al=-1,ar=-1;
        go(0,n.size()-1,&al,&ar);
        vector<int> ans;
        ans.push_back(al);
        ans.push_back(ar);
        return ans;
    }
};

File: Java/034. Search for a Range.java
/*
034. Search for a Range
https://leetcode.com/problems/search-for-a-range/

Java 8 ms
*/

public class Solution {
	// offset：-1表示找最左边的target，1表示找最右边的
	int bisearch(int nums[],int target,int offset)
	{
		int st=0,ed=nums.length-1,mid=0;
		while (st<=ed)
		{
			mid=(st+ed)/2;
			if (nums[mid]<target)
				st=mid+1;
			else if (nums[mid]>target)
				ed=mid-1;
			else
			{
				int pos=mid+offset;
				// 如果在offset一侧还有target，则继续找
				if (0<=pos&&pos<nums.length&&nums[pos]==target)
				{
					if (offset==-1)
						ed=mid-1;
					else
						st=mid+1;
				}
				else
					break;
			}
		}
		if (st<=ed)
			return mid;
		else
			return -1;
	}
	public int[] searchRange(int[] nums, int target) {
		int ret[]=new int[2];
		ret[0]=bisearch(nums,target,-1);
		ret[1]=bisearch(nums,target,1);
		return ret;
	}
}
